modepy/quadrature/finding.py

```diff
@@ -0,0 +1,422 @@
+        ) -> Sequence[LinearCombinationIntegrand]:
+    r"""
+    Let :math:`\Omega\subset\mathbb C^n` be a domain. (Domains
+    over the reals are allowable as well.) Returns linear combinations
+    of functions in *basis* that is orthogonal under the (complex-valued)
+    :math:`L^2` inner product induced by *integrate*.
+
+    :arg integrate: Computes an integral of the passed integrand over
+        :math:`\Omega`. Must use integration nodes compatible with
+        *basis*.
+    :arg basis: A sequence of functions that accept an array of nodes
+        of shape either ``(ndim, nnodes)`` or ``(nnodes,)`` and return
+        an array of shape ``(nnodes,)`` with the value of the
+        basis function at the node.
+    """
+    n = len(basis)
+    mass_mat = _mass_matrix(integrate, basis)
+
+    l_factor = la.cholesky(mass_mat)
+
+    from scipy.linalg import solve_triangular
+    l_inv = solve_triangular(l_factor, np.eye(n), lower=True)
+
+    assert la.norm(np.triu(l_inv, 1), "fro") < 1e-14
+
+    return [
+        linearly_combine(l_inv[i, :i+1], basis[:i+1])
+        for i in range(n)
+    ]
+
+
+@dataclass(frozen=True)
+class _ComplexToNDAdapter:
+    function: Integrand
+
+    def __call__(self, points: np.ndarray) -> np.ndarray:
+        rpoints  = np.array([points.real, points.imag])
+        return self.function(rpoints)
+
+
+def adapt_2d_integrands_to_complex_arg(
+            functions: Sequence[Integrand]
+        ) -> Sequence[Integrand]:
+    r"""Converts a list of :data:`Integrand`\ s taking nodes in real-valued
+    arrays of shape ``(2, nnodes)`` to ones accepting a single
+    complex-valued array of shape ``(nnodes,)``. See :func:`guess_nodes_vr`
+    for the main intended use case.
+    """
+    return [_ComplexToNDAdapter(f) for f in functions]
+
+
+def guess_nodes_vr(
+            integrate: Callable[[Integrand], np.inexact],
+            onb: Sequence[Integrand],
+        ) -> np.ndarray:
+    r"""
+    Let :math:`\Omega\subset\mathbb C` be a convex domain.
+    Finds interpolation nodes for :math:`\Omega` based on the
+    multiplication-operator technique in
+    [Vioreanu2011]_. May be useful as an initial guess for a Gauss-Newton
+    process to find new quadrature rules, as driven by
+    :func:`quad_residual_and_jacobian`.
+
+    :arg integrate: must accurately integrate a product of two functions from *onb* and
+        a degree-1 monomial over :math:`\Omega`.
+    :arg onb: An orthonormal basis of functions. Each function takes
+        in an array of (complex) nodes and returns an array
+        of the same shape containing the function values.
+        Functions are expected to be orthogonal with respect to
+        the complex-valued inner product.
+
+        Integrands accepting real-valued node coordinates in two dimensions
+        (shaped ``(2, nnodes)`` may be converted to functions acceptable by
+        this function via :func:`adapt_2d_integrands_to_complex_arg`.)
```

Comment:
Maybe worth adding a `is_real` flag or something to this to avoid exposing the complex adapter stuff?

Reply:
I thought about this, but ultimately the inner product (i.e. the integration) fundamentally happens in the complex domain, and the integrands are allowed to evaluate to complex numbers (even if they don't for most use cases I have in mind). As a result, I figured that something is genuinely lost by plastering a pretend-real-valued interface over this. And, just as a matter of taste, I strongly prefer composition with an additional function over passing a vaguely opaque flag (easier to understand, for one, and not too much more typing).

I guess I could see a pretend-real *wrapper*? Let me try that out.

Reply:
Tried the wrapper. Not worth it IMO. To be useful, it would have to deal with `(nnodes,)` and `(2, nnodes)`, and this caused enough grief to make me want to run away. So nah, not happening, sorry!